<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>HPC-REDItools - Guide</title>
    <link rel="stylesheet" type="text/css" href="../static/styles.css">
    <link rel="icon" type="image/x-icon" href="../static/logo.ico">
    <script src="../static/scripts.js" defer></script>
</head>
<body>

    <nav>
        <ul>
            <li><a href="index.html">Home</a></li>
            <li><a href="start.html">Start</a></li>
            <li><a href="download.html">Download</a></li>
            <li><a href="architecture.html">Architecture</a></li>
            <li><a href="help.html">Help</a></li>
            <li><a href="contacts.html">Contacts</a></li>
        </ul>
    </nav>

<div class="container">
    <div class="title">
        <div class="best" style="display: flex; align-items: center;">
            <img src="../static/logo.png" width="100" style="margin: auto;">
            <h1>HPC-T-Annotator</h1>
            <img src="../static/logo.png" width="100" style="margin: auto;">
        </div>
        <h2 id="subtitle">Interface Guide</h2>
    </div>

    <div class="description">
        <p>Welcome to the user guide for the HPC-T-Annotator interface. This guide will help you understand and utilize the features of the interface to configure and generate a customized software package for parallelization tasks involving annotation software. Follow the steps below to effectively use the interface.</p>
    </div>
    <div class="block">
        <h3>Step 1: Workload Manager Settings</h3>
        <div>
            <p>Select a workload manager to configure settings:</p>

            <ul>
                <li>
                    <input type="radio" id="slurm" value="slurm" name="workload_manager" onchange="toggleConfigurationPanel()" onload="toggleConfigurationPanel()">
                    <label for="slurm"><strong>Slurm</strong></label>
                </li>
                <li>
                    <input type="radio" id="htcondor" value="htcondor" name="workload_manager" checked onchange="toggleConfigurationPanel()" onload="toggleConfigurationPanel()">
                    <label for="HTC"><strong>HTCondor</strong></label>
                </li>
                <li>
                    <input type="radio" id="none" value="none" name="workload_manager" checked onchange="toggleConfigurationPanel()" onload="toggleConfigurationPanel()">
                    <label for="none"><strong>None</strong></label>
                </li>
            </ul>
        </div>

        <p>Based on your selection, different configuration options will be displayed.</p>

        <!-- Slurm configuration settings description here -->
        <div class="conf" id="slurmConfPanel" style="display: none;">
            <h4>Slurm Configuration Settings</h4>
            <p>Configure the settings for Slurm workload manager:</p>
            <ul>
                <li><strong>Account name</strong>: Specify the name of the account for job submission.</li>
                <li><strong>Serial partition</strong>: Enter the name of the serial partition for single-threaded jobs.</li>
                <li><strong>Parallel partition</strong>: Provide the name of the parallel partition for multi-threaded jobs.</li>
                <li><strong>Number of threads</strong>: Define the number of threads per process.</li>
                <li><strong>Wall-time(hours)</strong>: Specify the maximum duration of the job in hours.</li>
                <li><strong>Memory per process:</strong> Enter the memory requirement per process.</li>
            </ul>
        </div>

        <!-- Slurm configuration settings description here -->
        <div class="conf" id="htcondorConfPanel" style="display: none;">
            <h4>Slurm Configuration Settings</h4>
            <p>Configure the settings for Slurm workload manager:</p>
            <ul>
                <li><strong>Universe name</strong>: Specify the name of the HTCondor universe for job submission.</li>
                <li><strong>Schduler name</strong>: Enter the name of the HTCondor scheduler.</li>
                <li><strong>Accounting group</strong>: Provide the name of the accounting group for this job submission.</li>
                <li><strong>Number of threads</strong>: Define the number of threads per process.</li>
                <li><strong>Wall-time(hours)</strong>: Specify the maximum duration of the job in hours.</li>
                <li><strong>Memory per process:</strong> Enter the memory requirement per process.</li>
            </ul>
        </div>

        <!-- Base configuration settings description here -->
        <div class="conf" id="noneConfPanel">
            <h4>Base Configuration Settings</h4>
            <p>Configure basic settings when no workload manager is selected:</p>
            <ul>
                <li><strong>Number of processes</strong>: Set the desired number of processes.</li>
                <li><strong>Threads</strong>: Specify the number of threads for processing.</li>
            </ul>
        </div>

    </div>
    <div class="block">
        <h3>Step 2: REDItools Software Settings</h3>
        <div>
            <p>Configure the mandatory settings:</p>
        </div>
        <ul>
            <li><strong>Input directory absolute path</strong>: Insert the absolute path of the directory containing the input BAM files.</li>
            <li><strong>Absolute path to the genome reference</strong>: Insert the absolute path of the genome reference used to generate the BAM files.</li>
        </ul>

        <div>
            <p>Configure the optional settings:</p>
        </div>
        <ul>
            <li><strong>Absolute path of BED file containing target regions</strong>: Insert the absolute path of the BED file containing target regions</li>
            <li><strong>Absolute path to the file containing the splicing sites positions</strong>: Insert the absolute path of the file containing the splicing sites positions</li>
            <li><strong>Absolute path to the file containing the omopolymeric positions</strong>: Insert the absolute path of the file containing the omopolymeric positions</li>
            <li><strong>Additional options</strong>: Provide any additional command-line options or parameters for REDItools-v3 software.</li>
        </ul>

        <p>In note, for <strong>Additional options</strong> field, only the following options are accepted:</p>
        <ul>
            <li><strong>-s, --strand (choices: 0, 1, 2)</strong>: Specifies the strand orientation. It can be 0 (unstranded), 1 (second strand oriented), or 2 (first strand oriented).</li>
            <li><strong>-a, --append-file (action='store_true')</strong>: Enables the option to append results to an existing file (and creates it if it does not exist).</li>
            <li><strong>-g, --region</strong>: Enter the region of the BAM file to be analyzed.</li>
            <li><strong>-c, --create-omopolymeric-file (default=False, action='store_true')</strong>: Enables the option to create a file of omopolymeric positions.</li>
            <li><strong>-os, --omopolymeric-span (type=int, default=5)</strong>: Specifies the omopolymeric span.</li>
            <li><strong>-ss, --splicing-span (type=int, default=4)</strong>: Specifies the splicing span.</li>
            <li><strong>-mrl, --min-read-length (type=int, default=30)</strong>: Reads with a length below this value will be discarded.</li>
            <li><strong>-q, --min-read-quality (type=int, default=20)</strong>: Reads with mapping quality below this value will be discarded.</li>
            <li><strong>-bq, --min-base-quality (type=int, default=30)</strong>: Bases with quality below this value will not be included in the analysis.</li>
            <li><strong>-mbp, --min-base-position (type=int, default=0)</strong>: Bases residing in a previous position (in the read) will not be included in the analysis.</li>
            <li><strong>-Mbp, --max-base-position (type=int, default=0)</strong>: Bases residing in a further position (in the read) will not be included in the analysis.</li>
            <li><strong>-l, --min-column-length (type=int, default=1)</strong>: Positions whose columns have a length below this value will not be included in the analysis.</li>
            <li><strong>-men, --min-edits-per-nucleotide (type=int, default=0)</strong>: Positions whose columns have bases with fewer edits than this threshold will not be included in the analysis.</li>
            <li><strong>-me, --min-edits (type=int, default=0)</strong>: The minimum number of editing events (per position). Positions whose columns have bases with fewer edits than "min-edits-per-base edits" will not be included in the analysis.</li>
            <li><strong>-Men, --max-editing-nucleotides (type=int, default=100)</strong>: The maximum number of editing nucleotides, from 0 to 4 (per position). Positions whose columns have more than "max-editing-nucleotides" will not be included in the analysis.</li>
            <li><strong>-d, --debug (default=False, action='store_true')</strong>: Runs REDItools in DEBUG mode.</li>
            <li><strong>-T, --strand-confidence-threshold (type=float, default=0.7)</strong>: Reports strand orientation only if at least this proportion of reads are of a given strand.</li>
            <li><strong>-C, --strand-correction (default=False, action='store_true')</strong>: Enables strand correction. Once the strand has been inferred, only bases according to this strand will be selected.</li>
            <li><strong>-V, --verbose (default=False, action='store_true')</strong>: Provides verbose information on stderr.</li>
            <li><strong>-N, --dna (default=False, action='store_true')</strong>: Runs REDItools 2.0 on DNA-Seq data.</li>
            <li><strong>-w, --window (type=int, default=0)</strong>: Specifies how many base pairs should be processed by each thread at a time. Defaults to processing the entire contig.</li>
        </ul>

    </div>
    <div class="block">
        <h3>Generating the Software Package</h3>
        <div>
            <p>After configuring the settings for both workload manager and REDItools software, you can:</p>

            <ul>
                <li>Click the <strong>Reset the Fields</strong> button to clear all entered data.</li>
                <li>Click the <strong>Generate</strong> button to create the customized software package.</li>
            </ul>
        </div>
    

        <p>Once generated, you can download the software package in TAR format.</p>
    </div>
    <div class="description">  
        <p style="text-align: center;">The following step must be done on the HPC cluster.</p>
    </div>
    <div class="block">
        <h3>Extract the TAR archive and run</h3>
        <div>
            <p>After downloading the TAR archive (and, if you want, uploading the archive on HPC cluster) you have to unTAR the archive and then run the start 
               script.
            </p>
        </div>
        <div class="code-cell">
            tar -xf hpc-reditools.tar -C /path/to/myfolder/hpc-reditools<br>
            cd /path/to/myfolder/hpc-reditools
        </div>
        <p>And the, if you are on HPC cluster with Slurm workload manager</p>
        <div class="code-cell">
            sbatch start.sh
        </div>
        <p>Else, if you are on workload manager less architecture</p>
        <div class="code-cell">
            bash start.sh
        </div>
    </div>
    <div class="description">  
        <p style="text-align: center;">Once the entire computation process has ended (check the general.log file for the status), the final results will be in tmp/*.tsv files.</p>
    </div>
    <div class="block">
        <h3>Some useful monitoring commands</h3>
        <div>
            <h4>Check that all jobs are finished</h4>
        </div>
        <p>If all jobs are finished, you can check the logs by running this command:</p>
        <div class="code-cell">
            cat general.log
        </div>
        <p>That will display something like this:</p>
        <div class="code-cell">
            Starting timestamp#2023-09-16 14:03:14<br>
            Input files path: /g100_scratch/userexternal/larcioni/BAM<br>
            Processes: 24<br>
            Reference: /g100_scratch/userexternal/larcioni/test_mm10_bam/ref/genome.fa<br>
            Chromosomes: 22<br>
            Average runtime: 0:08:36<br><br>

            Max runtime: 0:12:08<br><br>

            Min runtime: 0:06:40<br><br>

            Ending timestamp#2023-09-16 14:30:58<br>
            Total elapsed time: 00:27:44<br>
        </div>
        <div>
            <h4>Check for errors</h4>
        </div>
        <p>You can check for jobs errors running these commands:</p>
        <div class="code-cell">
            # For general errors<br>
            cat general.err
        </div>
        <div class="code-cell">
            # For control script errors<br>
            cat control.err
        </div>
        <div class="code-cell">
            # For specific job errors<br>
            cat tmp/*/general.err
        </div>
    </div>
</div>
</body>
</html>

